packages/web/src/pages/campaign.js

```diff
@@ -0,0 +1,114 @@
+// Module imports
+import { useCallback, useEffect} from 'react'
+import { useRouter } from 'next/router'
+
+// Local imports
+import { Field } from 'components/Field'
+import { Form } from 'components/Form'
+import { FormButton } from 'components/FormButton'
+import { useAuth } from 'contexts/AuthContext'
+import {firestore, firebase} from '../helpers/firebase'
+
+
+export default function CampaignPage() {
+
+	const {
+		isLoggedIn
+		,user
+	} = useAuth()
+
+	const router = useRouter()
+	useEffect(() => 
+	{
+		if (!isLoggedIn)
+		{
+			router.push('/login');
+		}
+	})
+
+
+	const handleSubmit = useCallback(formState => {
+		const {
+			isValid,
+			values,
+		} = formState
+
+		if (isValid) {
+			ValidateAndSaveCampaign(values, user, router)
+		}
+	}, [ValidateAndSaveCampaign, user])
+
+	return (
+		<div>
+			<header><h2>Create Campaign</h2></header>
+			<Form 
+				initialValues={{
+					name: '',
+					description: ''
+				}}
+			onSubmit={handleSubmit}>
+			<Field
+					id="name"
+					isRequired
+					label="Name"
+					type="text"
+					maxLength={50} />
+
+				<Field
+					id="description"
+					isRequired
+					label="Description"
+					type="text"
+					maxLength={300} />
+
+				<FormButton
+					type="submit"
+					isDisabled={!isLoggedIn}>
+					Create
+				</FormButton>
+			</Form>
+		</div>
+		
+	)
+}
+
+function ValidateAndSaveCampaign(campaign, user, router)
+{
+	if (user.uid && campaign.description && campaign.name)
+	{
+		let isFirst = false;
+    	var campaignsCollection = firestore.collection('campaigns');
+		campaignsCollection
+			.where("ownerID", "==", user.uid)
+			.get()
+			.then((querySnapshot) => {
+				if (querySnapshot.empty)
+				{
+					isFirst = true;
+				}
+				
+				let newCampaign = campaignsCollection.doc()
+
+				newCampaign.set({
```

Comment:
You could shorten this...
```js
let newCampaign = campaignsCollection.doc()

newCampaign.set({})
```

...to a single line...

```js
let newCampaign = campaignsCollection.add({})
```

Reply:
I tried this and my later call to newCampaign.id returned undefined as a result. Is there a way to get a reference to the campaign ID when using add instead of my set pattern?
I want the campaign ID reference because I imagine passing it to the page we are routing so we can display the campaign.

Reply:
Yeah, the issue is that the document doesn't get an ID until after you save it, so if you `await` the `.add()`, you can pull the ID off of the result:

```js
const newCampaign = await campaignsCollection.add({})
console.log({
  id: newCampaign.id,
  data: newCampaign.data(),
})
```

* Reference: https://firebase.google.com/docs/firestore/manage-data/add-data#add_a_document

Reply:
I ended up pretty heavily refactoring this into modularized functions for saving/creating new campaigns.